src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs

```diff
@@ -0,0 +1,207 @@
+using System.Threading.Tasks;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.Execution;
+using Microsoft.Build.FileSystem;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Shared.FileSystem;
+
+namespace Microsoft.Build.Experimental.ProjectCache
+{
+    internal class ProjectCacheService
+    {
+        private readonly BuildManager _buildManager;
+        private readonly PluginLoggerBase _logger;
+        private readonly ProjectCacheDescriptor _projectCacheDescriptor;
+        private readonly CancellationToken _cancellationToken;
+        private readonly ProjectCacheBase _projectCachePlugin;
+
+        private ProjectCacheService(
+            ProjectCacheBase projectCachePlugin,
+            BuildManager buildManager,
+            PluginLoggerBase logger,
+            ProjectCacheDescriptor projectCacheDescriptor,
+            CancellationToken cancellationToken)
+        {
+            _projectCachePlugin = projectCachePlugin;
+            _buildManager = buildManager;
+            _logger = logger;
+            _projectCacheDescriptor = projectCacheDescriptor;
+            _cancellationToken = cancellationToken;
+        }
+
+        public static async Task<ProjectCacheService> FromDescriptorAsync(
+            ProjectCacheDescriptor pluginDescriptor,
+            BuildManager buildManager,
+            ILoggingService loggingService,
+            CancellationToken cancellationToken)
+        {
+            var plugin = await Task.Run(() => LoadPluginFromAssembly(pluginDescriptor.PluginPath), cancellationToken)
+                .ConfigureAwait(false);
+
+            // TODO: Detect and use the highest verbosity from all the user defined loggers. That's tricky because right now we can't discern between user set loggers and msbuild's internally added loggers.
+            var logger = new LoggingServiceToPluginLoggerAdapter(LoggerVerbosity.Normal, loggingService);
+
+            await plugin.BeginBuildAsync(
+                new CacheContext(
+                    pluginDescriptor.PluginSettings,
+                    new IFileSystemAdapter(FileSystems.Default),
+                    pluginDescriptor.ProjectGraph,
+                    pluginDescriptor.EntryPoints),
+                // TODO: Detect verbosity from logging service.
+                logger,
+                cancellationToken);
+
+            if (logger.HasLoggedErrors)
+            {
+                throw new Exception(
+                    ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("ProjectCacheInitializationFailed"));
+            }
+
+            return new ProjectCacheService(plugin, buildManager, logger, pluginDescriptor, cancellationToken);
+        }
+
+        private static ProjectCacheBase LoadPluginFromAssembly(string pluginAssemblyPath)
+        {
+            var assembly = LoadAssembly(pluginAssemblyPath);
+
+            var pluginType = GetTypes<ProjectCacheBase>(assembly).First();
```

Comment:
This will throw if there are no types that inherit from ProjectCacheBase, we'll probably want to detect that and log an error rather than an unhandled exception

Reply:
Fixed it here https://github.com/dotnet/msbuild/pull/5936/files#diff-ac47163cf1a4ff2fbd79d56e0cf79ce9636b6a8331dfa89dae3fe8d3547f7fb3R109